### PR TITLE
Fix cmd+click find all references fallback not working in Vim mode

### LINKED ISSUE #10392: `cmd+click` symbol definition to find all references does not work in vim mode
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When `"vim_mode": true`, `cmd + click` only `GoToDefinition` but does not `FindAllReference` when clicking on def (in INSERT or NORMAL mode).

If vim mode is disabled, however, the bidirectional `cmd + click` works properly.

https://github.com/zed-industries/zed/assets/52687642/22a71a2c-eb77-4804-a1ec-254e05e176fc

__

### Environment

Zed: v0.131.1 (Zed Preview)
OS: macOS 14.4.1
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

Bidirectional `cmd + click` should work also in vim mode. Related #9259 .

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

_No response_

### PR DESCRIPTION
**Pull request by Congyuwang**
Exclude go-to-definition links returned by LSP that points to the current cursor position. This fixes #10392 . Related PR #9243 .

The previous implementation first performs go-to-definition, and if the selected text covers the clicked position, it figures out that it is already clicking on a definition, and should instead look for references.

However, the selected range does not necessarily cover the clicked position after clicking on a definition, as in VIM mode.

After the PR, if cmd+click on definitions, the definition links would be an empty list, so no go-to-definition is performed, and find-all-references is performed instead.

Release Notes:

- vim: Fixed `cmd+click`ing on a definition to find all references #10392

### CODE REVIEW

**Comment by Congyuwang** [comment]
@cla-bot check

**Review by maxdeviant — CHANGES_REQUESTED** [review]
If I'm understanding this feature correctly, we had this functionality previously and reverted it due to both issues in the implementation and not being satisfied with the UX.

I think we'll need to evaluate this closely to make sure the behavior is exactly what we want.

Here is the other PR: https://github.com/zed-industries/zed/pull/6924
And the PR that reverted it: https://github.com/zed-industries/zed/pull/10094

Apologies if the two are unrelated, they just read as very similar to me.

**Comment by Congyuwang** [comment]
There is one side effect though. Now with cmd + cursor hovering over a definition, there would be no underline shown.

**Comment by Congyuwang** [comment]
> If I'm understanding this feature correctly, we had this functionality previously and reverted it due to both issues in the implementation and not being satisfied with the UX.
> 
> I think we'll need to evaluate this closely to make sure the behavior is exactly what we want.
> 
> Here is the other PR: #6924 And the PR that reverted it: #10094
> 
> Apologies if the two are unrelated, they just read as very similar to me.

I think this PR only wishes to fix previous #9243 (not reverted) not working in VIM mode.

**Comment by Congyuwang** [comment]
This really adds no feature, but is just a bug fix.

File: crates/project/src/project.rs [hunk 1]
```diff
@@ -4937,6 +4937,23 @@ impl Project {
...
+        // Exclude definition links that points back to cursor position.
+        // (i.e., currently cursor upon definition).
+        !(buffer == &location.target.buffer
+            && current_position
+                .cmp(&location.target.range.start, &snapshot)
+                .is_ge()
+            && current_position
+                .cmp(&location.target.range.end, &snapshot)
+                .is_le())
+    }
```
Comment (Congyuwang) [inline crates/project/src/project.rs:4955]: Could use `location.origin` to simplify this filter here, but `location.origin` is optional, and I'm not sure what `None` means for it.

File: crates/editor/src/hover_links.rs [hunk 2]
```diff
@@ -132,28 +131,12 @@ impl Editor {
...
-        };
-        drop(multi_buffer_snapshot);
-
         let reveal_task = self.cmd_click_reveal_task(point, modifiers, cx);
         cx.spawn(|editor, mut cx| async move {
             let definition_revealed = reveal_task.await.log_err().unwrap_or(false);
             let find_references = editor
                 .update(&mut cx, |editor, cx| {
-                    if definition_revealed && revealed_elsewhere(editor, before_revealing, cx) {
+                    if definition_revealed {
```
Comment (Congyuwang) [inline crates/editor/src/hover_links.rs:139]: Now, we are always "revealed elsewhere".

**Comment by ConradIrwin** [comment]
@Congyuwang Thanks for this, and sorry for the slow reply, it took a while to get familiar with the codepaths here.

Two changes needed to get this merged:
* Currently we filter links before showing them, however given that clicking on the definition does do something, it should be underlined. I think the best way to do this would be to remove the filter from `show_link_definition` and do it in `cmd_click_reveal_task` before passing the links in to navigate_to_hover_links.
* As `exclude_link_to_position` is only used by the hover links code, it should not be defined on Project, but in `hover_links.rs`. Secondly, as you are taking a `Model<Buffer>` as the first argument, you don't need to pass the snapshot, instead pass `cx` and then you can get one with `let snapshot = buffer.read(cx).snapshot(cx)`.

Let me know if that works out, or if there are further complexities I missed. Happy to pair on this if you'd like https://calendly.com/conradirwin/pairing

**Comment by Congyuwang** [comment]
@ConradIrwin Great. I see what you mean. It seems to work after the changes above, and the definition underlines are back. Also, Project is not touched now.

File: crates/editor/src/hover_links.rs [hunk 3]
```diff
@@ -180,12 +181,30 @@ impl Editor {
...
 
-                return self.navigate_to_hover_links(
-                    None,
-                    hovered_link_state.links,
-                    modifiers.alt,
-                    cx,
-                );
+                // exclude links pointing back to the current anchor
+                let current_position = point
+                    .next_valid
```
Comment (Congyuwang) [inline crates/editor/src/hover_links.rs:186]: I'm not sure whether `previous_valid` or `next_valid` should be used.

Comment (Congyuwang) [reply crates/editor/src/hover_links.rs:187]: Looks like they are the same after trying both.

Comment (ConradIrwin) [reply crates/editor/src/hover_links.rs:187]: They would only be different if you clicked on an inlay hint or similar. It's probably fine to just use one.

**Pull request merged**